base/lock.jl

```diff
@@ -556,3 +556,59 @@ end
     import .Base: Event
     export Event
 end
+
+const _NOT_CALLED = Ptr{Nothing}(0)
+const _STARTED = Ptr{Nothing}(1)
+const _DONE = Ptr{Nothing}(2)
+
+"""
+    Once()
+
+Create a `Once` object that can be used to execute a callback exactly once across all executions of the program.
+The callback is executed by calling `call_once!`. This is similar to `pthread_once` in POSIX.
+
+```
+    const once = Once()
+    foo() = println("This will be printed exactly once")
+    call_once!(once, foo)
+    call_once!(once, foo) # does nothing
+```
+
+!!! compat "Julia 1.12"
+    This functionality requires at least Julia 1.12
+"""
+mutable struct Once
+    @atomic done::Ptr{Nothing} # Ptr is used so it resets during precompilation
```

Comment:
Off. Does this need to be the semantics? Remember precompilation is an optimization,  normally Julia would execute the top-level code in the same session and precompilation is just "in the infinentisimal time window before init".

Now the argument for it resetting during precompilation would be that I am guarding a data structure that is being reset as well, but one might be making the opposing statement as well.

I kinda want to discourage people from running their precompilation workload inside the module while it is still open, rather we should run it outside the module and save the state as if the precompilation workload hasn't run yet (Nathan opened a PR, iirc)

Reply:
So there's two use cases here. One where I think this should just error if hit during precompilation. But you might be running a precompile workload that needs a library to be loaded/initialized and the only thing that would have the right semantics here is `Ptr`

Reply:
I think one might reasonably want both behavior `Once` and `OncePtr`. The challenge is that currently we don't know what the user is referring to. This kinda makes me wonder if we want this to be a ref to the data produces by F

Reply:
So the scenario I am thinking about is:

```julia
mutable struct Cache
    once::Once
    data
end
```

When Cache is being serialized it depends on what `data` is if I want to reset `once` or not. If data does not contain foreign pointers then `once` ought to not be reset.

Precompilation is equivalent to top-level execution in the target process, but we cache the effect. So IMO `Once` ought to serialized like any other Julia struct and have a reset function that can be used to reset it at the end of precompilation if it is guarding a foreign pointer.

Reply:
+1  that this is the really semantically trick part

> When Cache is being serialized it depends on what data is if I want to reset once or not. If data does not contain foreign pointers then once ought to not be reset.

I think the rule is more complicated than that... In many cases `Once` is intended to replace `__init__` functionality, in which case you might have cache sizes, local library paths, etc. which are selected - not just foreign pointers

None of those are supposed to affect the outcome of the computations at pre-compilation (even though they are generally different than what you'd compute with at run-time) so they are "sound" to compute twice, but you have to re-compute them again after pre-compilation or they're no longer valid.

Reply:
Yeah. I think have a `reset!` for `Once` would make sense so that the user can choose their behavior. It is IMO easier to explain what to do when your once computation needs to be to reset due to being run during precompilation, than to tell someone "it ran twice since we didn't think you would notice the first time" 

But that is also the reason why `__init__` is a separate thing.

Reply:
Hmm... Yeah, that's reasonable

Although, where would you put the `reset!` for a sysimage workflow? It'd have to go after all of your dependers in the sysimage, right? It's also a bit awkward that we already do some pretty terrible mutation teraring, since in the pkgimage workflow you get an automatic "reset" from us dropping mutations on the flow when you are run during pre-compilation for another package.

Oh, and the sysimage workflow is unfortunately more important now than before, since it's required for #55047.

I'd love to unify sysimage and pkgimage workflows semantically, but that seems a long way off still